### PR TITLE
Code review fix

### LINKED ISSUE #78: Track 1: NaN-fallback mechanism is structurally unusable outside the legacy segments path
From a full-repo review (line refs from `ecb7fe0`, pre-#77; may shift slightly after that PR merges).

PR #77 made the bundled models fail loud instead of silently zero-filling, but the fallback *mechanism* itself remains unusable for everything except WBMProbe:

1. **`LinearProbe._as_features` launders NaN away** — `src/openmhc/_probe.py:60-63` does `np.nan_to_num(x, nan=0.0, posinf=0.0, neginf=0.0)`. The documented external-submission pattern (custom encoder + LinearProbe) therefore *cannot* signal "cannot predict": NaN embeddings become zero vectors and produce finite predictions. External `fallback_rate` on the leaderboard is structurally 0.
2. **`_apply_fallback` requires the legacy segments path** — `src/downstream_evaluation/evaluation/evaluator.py:241-251`: when a method declares `needs_segments=False` (every bundled model except WBMProbe) or is predict-only, any non-finite prediction raises `NotImplementedError` instead of triggering the Linear fallback. Even a perfectly conformant NaN-emitting model can't be handled.
3. **The NaN convention is not stated in the public protocols** — the `Forecaster` protocol in `src/openmhc/_protocols.py` states "emit NaN where you cannot predict" explicitly; the `Method` protocol (`:81-152`) and the `Imputer` protocol (`:156-236`) say nothing. For Track 2 the contract exists only in the internal `imputation_evaluation/methods/base.py:60-66`. External submitters have no way to know NaN is the correct signal.

**Impact:** `fallback_rate` / `n_fallback` in Track 1 results and substrate sidecars is structurally 0 for all models except WBMProbe — not comparable to Track 2/3 where the rate is genuinely measured. Related: #39 (surfacing the rates on the leaderboard presumes they're meaningful).

**Suggested fix order:**
- Teach `_apply_fallback` to work without pre-supplied train segments (fetch them itself, or accept a fitted Linear fallback), removing the `NotImplementedError` path.
- Make `LinearProbe._as_features` propagate row-wise all-NaN embeddings to NaN predictions instead of zero-filling.
- Copy the Forecaster NaN wording into the `Method` and `Imputer` protocol docstrings.

### LINKED ISSUE #97: Umbrella: documentation drift (CLAUDE.md, README, submission packet, docstrings)
From a full-repo review (post-#77 state; #77 already fixes the C2/label_validity_criterion doc claim; #29 covers the paper_results README separately):

- **CLAUDE.md**: says 33 tasks (registry has 32); claims a `~/.cache/openmhc/data` default resolution tier (code deliberately raises — "Never falls back"); references `scripts/downstream_eval/` and `scripts/run_forecasting_eval.py` (neither exists — actual surfaces are `scripts/run_eval.py` and the `mhc-forecast-eval` / `mhc-impute-eval` console scripts); says E501 is disabled "for docstrings" (it's global).
- **README.md:138**: `mhc-downstream-eval --multirun method=linear,mae,xgboost` — no `mae` method exists in downstream; the multirun errors.
- **`openmhc/_submission.py:151-154`**: forecasting submission packet still says "format being finalized" while `tools/leaderboard_docs/forecasting/SCHEMA.md` declares the format final.
- **`paper_metrics_core.py:429-437,637-641`**: docstrings promise an `overall_binary_collapsed` scope that is never emitted.
- **`evaluate_prediction` / `mhc-downstream-eval` `seed`**: documented knob never reaches the classifiers (registry builders don't forward it) — either wire it or document it as inert.
- **`models/lsm2/blocks.py:96-134`**: docstring says a `(B, N)` boolean "keep" mask is supported, but it's applied as an *additive bias* — a 1/0 keep mask becomes a +1.0 logit bias on kept keys, i.e. no masking at all (verified numerically). In-repo callers pass the 4D additive form so nothing internal breaks, but any external caller following the docstring gets silently wrong attention. Fix the docstring or convert boolean masks to -inf bias (related LSM2 attention context: #11).

### PR DESCRIPTION
**Pull request by Yuze-312**
Fixes the 6 correctness issues from the Track 1 review. Bug-for-bug below;
1. **Ordinal preds rounded to {0,1}** — `_combine_with_fallback` rank-merges ordinal
   predictions into (0,1], then the metric + the persisted `y_pred` both did
   `np.round().astype(int)`, collapsing every rank to a binary median-split (incl. the
   model's own finite preds) and propagating into the paper bootstrap + leaderboard.
   **Fix:** score & persist the raw ordinal prediction (Spearman is rank-based; the int
   cast now applies only to multiclass). `evaluator.py`, `predictions_io.py`.
2. **XGBoost FPCA/imputation fit on train+val+test** (transductive). **Fix:** fit on the
   train split only, transform all users (version-aware split lookup).
3. **GRU-D uses the test cohort + wrong lookup** — (a) the decay-to-mean `empirical_mean`
   was recomputed from the test cohort at predict; (b) it read the *windowed* label lookup
   while the engine evaluates full-history. **Fix:** cache the train-split mean and reuse it
   at val/predict; switch to the full-history lookup (matches the other models).
4. **5 models silently zero-fill missing features** → scored as real, bypassing the
   NaN→Linear fallback. **Fix:** the zero-fill is never triggered on the shipped cohort
   (verified: cohort ⊆ feature store for all), but replaced it with a fail-loud `ValueError`
   guard in each model so a future data mismatch can't silently score fabricated predictions.
5. **`np.clip` corrupts AUPRC for score/logit submissions** (ties everything at the bound;
   AUPRC is rank-based and finiteness is already guaranteed). **Fix:** delete the clip.
6. **C2 (`weekly_5of7`) / `label_validity_criterion` documented but unimplemented.**
   **Fix:** dropped C2 + the config promise from the docs; document only the shipped C1 rule.

Closes #78.

Addresses part of #97: fixes the CLAUDE.md task count/default-path/CLI/E501/validity docs, the README downstream multirun example, the `evaluate_prediction(seed=...)` wording, and the LSM2 attention-mask docstring. Does not close #97 because the umbrella still includes unresolved documentation-drift items outside this PR.

### CODE REVIEW

**Review by NarayanSchuetz — CHANGES_REQUESTED** [review]
## Code review — 8 finder angles, adversarially verified

Automated high-effort review of this PR (line-by-line scan, removed-behavior audit, cross-file trace, reuse/simplification/efficiency/altitude/conventions angles; every candidate independently verified against the code, refuted candidates dropped). The six fixes are directionally correct — and two "the new guards will crash on legitimate data" concerns were **refuted** during verification (TSFM and LSM2 cohort membership structurally implies a cached feature). 10 findings survived, most severe first.

---

### 1. CONFIRMED — The FPCA leak fix silently doesn't take effect on any machine with an existing feature cache
`src/downstream_evaluation/models/xgboost/model.py:212`

The train-only fit (issue 2 of this PR) runs only inside `if force or not ca_out.exists():`. A `pipeline_curve_analysis_user_features.parquet` built by the old transductive code is reused verbatim — same filename, no version stamp or provenance check, and `_ensure_features` (model.py:299–311) never passes `force`. Any host with a populated `results/xgboost_features/` cache keeps reporting test-leaked FPCA numbers while the code claims train-only fitting. Suggest bumping the cache filename (or writing/checking a sidecar marker) so stale leaky features can't be silently loaded.

### 2. CONFIRMED — Leaderboard schema docs now contradict the persisted ordinal `y_pred`
`tools/leaderboard_docs/downstream/SCHEMA.md:52`, `tools/leaderboard_docs/SCHEMA.md:20`

Both docs still say `y_pred` is "the model's discrete prediction (class for binary/ordinal/multiclass…)", while `write_task_predictions` now persists **continuous** ordinal `y_pred` — and `bootstrap_skill_rank` computes ordinal Spearman directly from that column. An external submitter following the doc ships rank-collapsed integer classes and is compared against raw-score substrates: incommensurable encodings on the public leaderboard. The docs need the same update as the code.

### 3. PLAUSIBLE — Mixed-generation prediction stores bias the paired bootstrap
`src/downstream_evaluation/evaluation/bootstrap_skill_rank.py:143`

Parquets written pre-PR store ordinal `y_pred` as `np.round(raw).astype(int64)` (collapsed to {0,1} for fallback-merged scores); post-PR parquets store raw floats. Nothing regenerates or version-stamps existing stores, and both encodings are plain numeric columns, so a paired comparison between a method evaluated before this PR and one evaluated after silently degrades the older method's Spearman. Existing ordinal substrates need regeneration (or a version stamp that makes mixing detectable).

### 4. CONFIRMED — GRU-D: the desync the new guard catches at predict is still silent through training
`src/downstream_evaluation/models/grud.py:378` vs `:467`

`_build_split` still silently drops cohort users lacking segments (`users = [u for u in sorted(cohort) if u in self._segments]`), while `predict` now raises for exactly that condition. On a genuine lookup/segment-store desync, the full multi-task GPU fit (up to 50 epochs) completes on a silently reduced train/val cohort — corrupting the model, the cached train `empirical_mean`, and early stopping — and only detonates at the first `predict`. The same fail-loud check belongs in `_build_split` (or at least in `fit`) so the mismatch is caught before hours of training.

### 5. PLAUSIBLE — GRU-D hard-codes the full-history lookup; windowed ablation would silently mismatch
`src/downstream_evaluation/models/grud.py:367`

`lookup_filename("daily", full_history=True)` fixes the default run (matching linear/lsm2/xgboost), but the engine selects its lookup from `config.temporal.is_full_history` (runner.py:47) and GRU-D receives no temporal config. Under `TemporalWindowConfig.windowed()` — currently reachable only programmatically — GRU-D would train on full-history cohorts/labels while being scored on windowed test cohorts, with no error. Worth a comment or an explicit `set_temporal_window` rejection so the ablation can't run silently wrong.

### 6. PLAUSIBLE — Train-restricted curve statistics have unguarded degenerate paths
`src/downstream_evaluation/models/xgboost/pipeline_curve_analysis.py:337` and `:234`

Empty `train_pos` makes `fd_valid.coefficients[train_pos].mean(axis=0)` all-NaN (RuntimeWarning only), silently written into every insufficient-obs user; and `_impute_population_mean(matrix, fit_mask)` now fills any minute-column that is all-NaN *among train users* with 0.0 (e.g. 0-bpm heart rate) even when held-out users have real observations there. Reachable with the xs dev subset, a custom splits file, or a stale `curve_analysis_avg_curves*.parquet` checkpoint whose users barely intersect the train split. A `fit_mask.any()` / non-empty `train_pos` assertion would keep this fail-loud like the rest of the PR.

### 7. PLAUSIBLE — WBM guard aborts the eval where the harness's designed fallback exists
`src/downstream_evaluation/models/wbm/model.py:281`

WBM is the one model where the per-user NaN→Linear fallback actually works (`needs_segments=True`, evaluator.py:215–228). The weekly labels lookup and the window-index-derived embedding cache are two independently shipped artifacts (plus a persistent on-disk embedding cache that can go stale), so a missing (user, week) key is realistic — and now kills the entire 33-task WBM eval inside `WBMProbe.fit`/`predict`. Raising beats the old silent zero-fill, but emitting NaN for the missing user would route it to the Linear fallback and preserve graceful degradation. Design call — flagging the tradeoff.

### 8. CONFIRMED — Both leak fixes keep the leaky behavior as the silent default
`src/downstream_evaluation/models/xgboost/pipeline_curve_analysis.py:505`, `src/downstream_evaluation/models/grud.py:54`

`build_curve_analysis_features(fit_user_ids=None)` is documented as "fits on every user (legacy)", and `_UserDataset(empirical_mean=None)` silently recomputes the fitted decay statistic on the current cohort. Every in-repo caller already passes the safe value, so making the parameters required costs nothing — while the `None` defaults mean a future/external caller silently reproduces the exact leakage this PR fixes. This also runs against the repo CLAUDE.md rules "No silent fallbacks" / "No backwards compatibility by default".

### 9. CONFIRMED — The fail-loud guard is copy-pasted five times; only one copy is tested
`multirocket.py:161–174`, `tsfm.py:455–470`, `lsm2/model.py:280–301`, `wbm/model.py:263–285`, `grud.py:456–473`

`tsfm._encode` and `multirocket._features` are line-for-line identical after the preamble, and the sole regression test covers MultiRocket only (its docstring admits it "stands in for the shared pattern"). A shared `aligned_feature_matrix(name, user_ids, by_user, store_desc)` helper (nothing suitable exists in `data/` or `models/` today) would make the tested implementation the only implementation and stop a 6th model from silently reintroducing zero-fill.

### 10. CONFIRMED — The per-task-type y_pred policy now lives in two ladders that must drift in lockstep
`src/downstream_evaluation/evaluation/predictions_io.py:68` and `evaluator.py:26–40`

This PR itself had to patch ordinal in both `_metrics_for` and `write_task_predictions` simultaneously (plus the implicit third copy in `bootstrap_skill_rank`'s per-type column choice), and no test cross-checks them. A future per-type change applied to one ladder makes the live metric and the persisted substrate silently disagree. A single `prepare(ttype, raw) -> (y_pred_col, y_proba)` helper consumed by both would remove the drift surface.

---

**Minor (below the cap):** the now-vacuous `if str(u) in self._segments` filter and derivable `pred_by_user` rebuild left in `grud.py:474/495` after the new guard; and `jobs/imperial/pbs/run_eval.pbs:27`, `run_eval_gpu.pbs:31`, `jobs/imperial/slurm/run_eval.slurm:41` still hard-code the `py310` env this PR migrates away from.

**Comment by NarayanSchuetz** [comment]
### 11. CONFIRMED — The demographic-aware Linear path still silently fabricates all-zero covariates
`src/downstream_evaluation/demo_covariates.py:84`, `:127`; called from `src/downstream_evaluation/models/linear.py:107`

This PR replaces silent zero-fill with fail-loud guards in the embedding/feature models, but the shipped Track 1 demographic-aware Linear baseline still has the same bug class through `apply_demographics`:

```python
demo_matrix = np.zeros((len(uids), len(cov_indices)), dtype=np.float32)
for row_idx, uid in enumerate(uids):
    vec = demo_user_lookup.get(uid)
    if vec is not None:
        demo_matrix[row_idx] = vec[cov_indices]
```

A cohort user absent from `demo_user_lookup` silently keeps an all-zero demographic row, so Linear can score fabricated `age=0`, `BiologicalSex=0`, `BMI=0` instead of failing loud. This is the same zero-fill/desync pattern the PR fixes in `multirocket`, `tsfm`, `lsm2`, `wbm`, and `grud`; it just missed the demographic covariate path.

There is a second, milder zero-fill in `build_demo_user_lookup_from_labels_df`:

```python
_user_demo = _demo_df.groupby("user_id")[demo_covariates].first().fillna(0.0)
```

That converts sentinel/missing demographic values to zero after `pd.to_numeric(..., errors="coerce")`. Unlike an absent lookup user, per-covariate imputation may be a modeling choice, but `age=0`/`BMI=0` are out of distribution and `BiologicalSex=0` can collide with a real encoded category. At minimum this should count/log imputed users or add an explicit missingness treatment; for the absent-user case, the consistency fix is the same fail-loud guard used elsewhere.

**Review by NarayanSchuetz — APPROVED** [review]
LGTM

**Comment by NarayanSchuetz** [comment]
### GitHub issue closure status

This PR closes **#78**: Track 1 NaN fallback is now usable outside the legacy WBM/segments path.

It also addresses part of **#97** by fixing several documentation-drift items touched by this PR, but **does not close #97** because that umbrella issue still has unresolved items outside this PR.

No other GitHub issues should be auto-closed by this PR.
